src/compiler.ts

```diff
@@ -2681,12 +2681,87 @@ export class Compiler extends DiagnosticEmitter {
   private compileForOfStatement(
     statement: ForOfStatement
   ): ExpressionRef {
-    this.error(
-      DiagnosticCode.Not_implemented_0,
-      statement.range,
-      "Iterators"
-    );
-    return this.module.unreachable();
+    const compiler = this;
+    function earlyBailout(ty: Type) {
+      compiler.error(
+        DiagnosticCode.Type_0_is_not_iterable,
+        statement.range,
+        ty.toString()
+      );
+      return compiler.module.unreachable();
+    }
+    if (statement.variable.kind != NodeKind.VARIABLE) {
+      this.error(
+        DiagnosticCode.Not_implemented_0,
+        statement.range,
+        "For .. of where the target is not a variable"
+      );
+      return this.module.unreachable();
+    }
+    const variable = <VariableStatement>statement.variable;
+    if (variable.declarations.length != 1) {
+      this.error(
+        DiagnosticCode.Only_a_single_variable_declaration_is_allowed_in_a_for_of_statement,
+        statement.range
+      );
+      return this.module.unreachable();
+    }
+    const iterable = this.compileExpression(statement.iterable, Type.auto);
+
+    const iterablety = this.resolver.resolveExpression(statement.iterable, this.currentFlow);
+    if (!iterablety) return this.module.unreachable();
+    const iterablecls = iterablety.getClass();
+    if (!iterablecls) return earlyBailout(iterablety);
+    const iterableelem = iterablecls.lookupIterator();
+    if (!iterableelem) return earlyBailout(iterablety);
+    const iterablefn = <Function>iterableelem;
+    const iterator = iterablefn.signature.returnType;
+    const getiteratorcall = this.compileCallDirect(iterablefn, [], statement.iterable, iterable);
+    const iteratortemp = this.currentFlow.getTempLocal(iterator);
+
+    const iteratorcls = iterator.getClass();
+    if (!iteratorcls) return earlyBailout(iterablety);
+    
+    const nextelem = iteratorcls.getMethod('next');
+    if (!nextelem) return earlyBailout(iterablety);
+
+    const doneelem = iteratorcls.getMethod('done');
+    if (!doneelem) return this.module.unreachable();
+    
+    const iteratorvaluetype = (<Function>nextelem).signature.returnType;
```

Comment:
The cast here (and some below) can fail when `nextelem` (or `doneelem`) is not a `Function`, which would lead to a compiler crash. To mitigate, the `kind` of the element can be checked explicitly, and a diagnostic produced if it is not of the required type or shape. With shape I mean that there also needs to be a signature check that calling the methods above will actually produce valid WebAssembly code, plus making sure that the methods are indeed compiled before being called.